onnx/defs/controlflow/defs.cc

```diff
@@ -0,0 +1,215 @@
+   the variables b and keepgoing) must be declared as either loop-carried
+   dependencies (both at the op inputs and output and at the body net input and
+   output) or scan_outputs.
+3) Values created in the body cannot be accessed in the enclosing scope.
+
+Note that the semantics of this op support "diagonal" or "wavefront" execution.
+(See Step 3 here for an example:
+https://devblogs.nvidia.com/optimizing-recurrent-neural-networks-cudnn-5/).
+Frontends should emit multi-layer RNNs as a series of While operators (with
+time being the inner looping dimension), with each successive layer consuming
+the scan_outputs from the previous layer, possibly going through several
+point-wise operators (e.g. dropout, residual connections, linear layer).
+Concretely, the (possibly transformed) scan_outputs are referenced by the
+subsequent layer as a LoopIndexTensor operating on a value in scope, not
+necessarily a loop-carried dependency. Backends can recognize this pattern and
+are permitted to schedule the execution of the multi-layer network in a
+pipelined/"wavefront" fashion.
+
+)DOC")
+    .Input(
+        0,
+        "M",
+        "A maximum trip-count for the loop specified at runtime. Optional."
+        " pass empty string to skip.",
+        "I")
+    .Input(
+        1,
+        "cond",
+        "A boolean termination condition. Pass empty string to skip.",
+        "B")
+    .Input(
+        2,
+        "v_initial",
+        "The initial values of any loop-carried dependencies (values that "
+        "change across loop iterations)",
+        "V",
+        OpSchema::Variadic)
+    .Output(
+        0,
+        "v_final_and_scan_outputs",
+        "Final N loop carried dependency values then K scan_outputs",
+        "V",
+        OpSchema::Variadic)
+    .Attr(
+        "body",
+        "The graph run each iteration. It has 2+N inputs: (iteration_num, "
+        "condition, loop carried dependencies...). It has 1+N+K outputs: "
+        "(condition, loop carried dependencies..., scan_outputs...). Each "
+        "scan_output is created by concatenating the value of the specified "
+        "output value at the end of each iteration of the loop. It is an error"
+        " if the dimensions of these values change across loop iterations.",
+        AttributeProto::GRAPH,
+        true)
+    .TypeConstraint("V", OpSchema::all_tensor_types, "All Tensor types")
+    .TypeConstraint("I", {"int64"}, "Only int64")
+    .TypeConstraint("B", {"bool"}, "Only bool");
+
+OPERATOR_SCHEMA(LoopIndexTensor)
+    .SetSupportLevel(SupportType::EXPERIMENTAL)
+    .SetDoc(
+        "This is a special operator only valid inside the loop that supports "
+        "the common case behavior of accessing the correct element of the input"
+        " sequence in an RNN. This operator MUST be directly given the passed-"
+        "in iteration number to the body of a Loop graph. This signals to back-"
```

Comment:
With the changes, this is no longer a special operator valid only inside a loop. I am not sure I understand the restrictions "MUST be ...". Some optimizations may be valid only under these conditions, but isn't that up to the compiler/runtime to figure out? I suggest changing the name to "IndexTensor", changing input-name "loop_idx" to "idx" and the documentation to say that the operation indexes the specified axis using the specified index-value to produce a sub-tensor.